File: programas/metodo-falsa-posicion/falsa.go
package main

import (
	"fmt"
	"math"
)

func main() {
	n := 13.0
	tol := 0.5 * math.Pow(10, 2 - n)
	root, eps_a := bisect(f, -3, -1, 0, tol, 0)
	fmt.Println(root, "\t", eps_a)
}

func bisect(f func(x float64) float64, low, up, prev, tol, depth float64) (float64, float64) {
	middle := up - (f(up) * (low - up)) / (f(low) - f(up))

	eps_a := math.Abs((prev - middle) / middle) * 100

	fmt.Printf("%v\t%v\t%v\n", depth, middle, eps_a)

	if depth > 100 || eps_a < tol {
		return middle, eps_a
	}

	if f(low) * f(middle) < 0 {
		return bisect(f, low, middle, middle, tol, depth + 1)
	} else {
		return bisect(f, middle, up, middle, tol, depth + 1)
	}
}

func f(x float64) float64 {
	return x*x - 4 - math.Sin(x)
}

File: programas/metodo-falsa-posicion/falsa.f90
PROGRAM biseccion
        IMPLICIT NONE
        REAL (KIND=16) root, bisect, eps_s
        INTEGER n

        n = 13
        eps_s = 0.5d0 * (10.d0 ** (2 - n))

        root = bisect(-1._16, 3._16, 0._16, eps_s, 0)
END PROGRAM biseccion

FUNCTION f(x)
        IMPLICIT NONE
        REAL (KIND=16) f, x

        f = (x**2) - 4.d0 - SIN(x)
END FUNCTION

RECURSIVE FUNCTION bisect(xl, xu, prev, eps_s, depth) RESULT(res)
        IMPLICIT NONE
        REAL (KIND=16) xl, xu, xr, res, prev, f, eps_s, eps_a
        INTEGER depth

        xr = xu - (f(xu) * (xl - xu)) / (f(xl) - f(xu))

        eps_a = ABS((xr - xl) / xr) * 100

        WRITE(*,*) depth, xr, eps_a
        IF (depth > 100 .or. eps_a < eps_s) THEN
                res = xr
                RETURN
        END IF

        IF (f(xl) * f(xr) < 0) THEN
                res = bisect(xl, xr, xr, eps_s, depth + 1)
        ELSE
                res = bisect(xr, xu, xr, eps_s, depth + 1)
        END IF

        res = 0
END FUNCTION
